File: web/src/lib/types/services/dynamic-dns.ts
import { z } from 'zod/v4';

export const DynamicDNSEntrySchema = z.object({
	id: z.number(),
	enabled: z.boolean(),
	provider: z.string(),
	providerSettings: z.record(z.string(), z.string()).nullish().transform((value) => value ?? {}),
	credentialConfigured: z.boolean(),
	hostname: z.string(),
	recordType: z.enum(['A', 'AAAA', 'BOTH']),
	intervalMinutes: z.number(),
	sourceType: z.enum(['interface', 'manual', 'stun']),
	sourceSettings: z.record(z.string(), z.string()).nullish().transform((value) => value ?? {}),
	lastStatus: z.string().default(''),
	lastError: z.string().default(''),
	ipv4Status: z.string().default(''),
	ipv4Error: z.string().default(''),
	ipv6Status: z.string().default(''),
	ipv6Error: z.string().default(''),
	lastIPv4: z.string().default(''),
	lastIPv6: z.string().default(''),
	lastSyncAt: z.string().nullable().optional(),
	lastSuccessAt: z.string().nullable().optional(),
	createdAt: z.string(),
	updatedAt: z.string()
});

export type DynamicDNSEntry = z.infer<typeof DynamicDNSEntrySchema>;

export interface DynamicDNSEntryInput {
	enabled: boolean;
	provider: 'cloudflare' | 'namecheap';
	providerSettings?: Record<string, string>;
	token?: string;
	hostname: string;
	recordType: 'A' | 'AAAA' | 'BOTH';
	intervalMinutes: number;
	sourceType: 'interface' | 'manual' | 'stun';
	sourceSettings: Record<string, string>;
}

File: internal/db/models/dynamicdns/dynamicdns.go
// SPDX-License-Identifier: BSD-2-Clause
//
// Copyright (c) 2025 The FreeBSD Foundation.
//
// This software was developed by Hayzam Sherif <hayzam@alchemilla.io>
// of Alchemilla Ventures Pvt. Ltd. <hello@alchemilla.io>,
// under sponsorship from the FreeBSD Foundation.

package dynamicdnsModels

import "time"

const (
	ProviderCloudflare = "cloudflare"
	ProviderNamecheap  = "namecheap"

	RecordTypeA    = "A"
	RecordTypeAAAA = "AAAA"
	RecordTypeBoth = "BOTH"

	SourceTypeInterface = "interface"
	SourceTypeManual    = "manual"
	SourceTypeSTUN      = "stun"
)

type Entry struct {
	ID uint `json:"id" gorm:"primaryKey;autoIncrement"`

	Enabled          bool              `json:"enabled" gorm:"not null;default:true"`
	Provider         string            `json:"provider" gorm:"not null;index"`
	ProviderSettings map[string]string `json:"providerSettings" gorm:"serializer:json;type:json"`
	ProviderSecret   string            `json:"-"`

	Hostname        string `json:"hostname" gorm:"not null;index"`
	RecordType      string `json:"recordType" gorm:"not null"`
	IntervalMinutes uint   `json:"intervalMinutes" gorm:"not null;default:10"`

	SourceType     string            `json:"sourceType" gorm:"not null"`
	SourceSettings map[string]string `json:"sourceSettings" gorm:"serializer:json;type:json"`

	LastStatus    string     `json:"lastStatus"`
	LastError     string     `json:"lastError"`
	IPv4Status    string     `json:"ipv4Status"`
	IPv4Error     string     `json:"ipv4Error"`
	IPv6Status    string     `json:"ipv6Status"`
	IPv6Error     string     `json:"ipv6Error"`
	LastIPv4      string     `json:"lastIPv4"`
	LastIPv6      string     `json:"lastIPv6"`
	LastSyncAt    *time.Time `json:"lastSyncAt"`
	LastSuccessAt *time.Time `json:"lastSuccessAt"`

	CreatedAt time.Time `json:"createdAt" gorm:"autoCreateTime"`
	UpdatedAt time.Time `json:"updatedAt" gorm:"autoUpdateTime"`
}

func (Entry) TableName() string {
	return "dynamic_dns_entries"
}
